web/src/main/java/com/epam/esm/web/controller/CertificateController.java

```diff
@@ -2,94 +2,130 @@
+import java.util.Collections;
 import java.util.List;
-import java.util.Map;
-import java.util.stream.Collectors;
+
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.methodOn;
 
 @RestController
 @RequestMapping(value = "/certificates")
 public class CertificateController {
-    private final TagServiceImpl tagServiceImpl;
-    private final CertificateServiceImpl certificateServiceImpl;
+    private final TagService tagService;
+    private final CertificateService certificateService;
+    private final TagDTOMapper tagDTOMapper;
+    private final CertificateDTOMapper certificateDTOMapper;
 
     @Autowired
-    public CertificateController(TagServiceImpl tagServiceImpl,
-                                 CertificateServiceImpl certificateServiceImpl) {
-        this.tagServiceImpl = tagServiceImpl;
-        this.certificateServiceImpl = certificateServiceImpl;
+    public CertificateController(TagService tagService,
+                                 CertificateService certificateService,
+                                 CertificateDTOMapper certificateDTOMapper,
+                                 TagDTOMapper tagDTOMapper) {
+        this.tagService = tagService;
+        this.certificateService = certificateService;
+        this.certificateDTOMapper = certificateDTOMapper;
+        this.tagDTOMapper = tagDTOMapper;
     }
 
     @GetMapping
-    public ResponseEntity<?> readAll(@RequestParam Map<String, String> params)
-            throws ServiceException, BadRequestException {
-        Collection<CertificateDTO> list;
-            if (CollectionUtils.isEmpty(params)) {
-                list = certificateServiceImpl.readAll()
-                        .stream().map(CertificateDTOConverter::convertObject).collect(Collectors.toList());
-            } else {
-                list = certificateServiceImpl.read(params)
-                        .stream().map(CertificateDTOConverter::convertObject).collect(Collectors.toList());
-            }
-        return !CollectionUtils.isEmpty(list)
-                ? new ResponseEntity<>(list, HttpStatus.OK)
-                : new ResponseEntity<>(HttpStatus.NOT_FOUND);
+    public ResponseEntity<?> read(@RequestParam MultiValueMap<String, String> params)
+            throws NotFoundException, BadRequestException {
+        List<Certificate> certificates;
+        if (CollectionUtils.isEmpty(params)) {
+            certificates = certificateService.readAll();
+        } else {
+            certificates = certificateService.read(params);
+        }
+        CollectionModel<CertificateDTO> certificateDTOs = certificateDTOMapper.toCertificateDTOList(
+                certificates);
+        Link link = linkTo(methodOn(this.getClass()).read(params)).withRel("certificates");
+        certificateDTOs.add(link);
+        Paginator paginator = new Paginator(params);
+
+        if(paginator.isLimited(params)) {
```

Comment:
this block of code is used in several places, so it's better to move it to a separate method

Reply:
added private methods in each controller